system/lib/wasmfs/wasmfs.cpp

```diff
@@ -157,18 +159,21 @@ __wasi_fd_t __syscall_open(long pathname, long flags, long mode) {
   pathParts.push_back(interim);
 
   std::shared_ptr<File> curr = getRootDirectory();
-  for (int i = 0; i < pathParts.size(); i++) {
+  for (int i = 0; i < pathParts.size() - 1; i++) {
+
+    auto file = curr->dynCast<Directory>();
 
-    // If it is a file and we have not reached the end of the path, exit
-    if (curr->is<DataFile>() && i != pathParts.size() - 1) {
+    // If file is nullptr, then the file was not a Directory.
+    // TODO: Change this to accommodate symlinks
+    if (!file) {
       return -(EBADF);
     }
 
     // Find the next entry in the current directory entry
 #ifdef WASMFS_DEBUG
-    curr->get().printKeys();
+    file->get().printKeys();
 #endif
-    curr = curr->cast<Directory>()->get().getEntry(pathParts[i]);
+    curr = file->get().getEntry(pathParts[i]);
```

Comment:
We should also handle (and test) the error case where `getEntry` returns null here because the requested directory does not exist.